terra/executor/resources.py

```diff
@@ -0,0 +1,395 @@
+'''
+In a multithreaded or multiprocessing environment, it may become necessary to
+divide up and balance a limited "resource" among multiple workers. When a
+:class:`terra.task.TerraTask` is distributed to a worker, each worker might
+need to know what resource id to use, when you have this limited resource.
+
+For example if you have 4 GPUs and each GPU can only handle two simultaneous
+workers at a time, you want to have a total of 8 workers, but at any one time,
+you want two workings working on a specific GPU.
+
+In order to balance this, a :class:`Resource` is maintained
+'''
+
+import os
+import threading
+import platform
+import tempfile
+import atexit
+import weakref
+from shutil import rmtree
+import vsi.vendored.filelock as filelock
+
+from vsi.tools.dir_util import is_dir_empty
+
+from terra import settings
+from terra.executor import Executor
+
+from terra.logger import getLogger
+logger = getLogger(__name__)
+
+
+class ResourceError(Exception):
+  pass
+
+
+class ProcessLocalStorage:
+  '''
+  Processor local storage class
+  '''
+  resource_id = None
+  instance_id = None
+  lock = None
+
+
+class ThreadLocalStorage(ProcessLocalStorage, threading.local):
+  '''
+  Thread version of :class:`ProcessLocalStorage`
+  '''
+
+
+class Resource:
+  '''
+  A :class:`Resource` instance will represent a set of resources that can be
+  acquired among multiple threads or processes.
+
+  This library is intended to be a coarse grain allocation, meaning it is
+  intended that acquiring a resource lasts the life of the thread/process, not
+  just for a function call or a section of a function. Multiple calls to
+  :meth:`acquire` or using ``with`` will result in the same resource, that was
```

Comment:
some other way to say this: that was not released between calls. also, if you did this, would the resource need to be released multiple times?